scylla-server-rust/src/main.rs

```diff
@@ -7,44 +7,153 @@ use scylla_server_rust::{
         run_controller, system_controller,
     },
     prisma::PrismaClient,
+    processors::{
+        db_handler, mock_processor::MockProcessor, mqtt_processor::MqttProcessor, ClientData,
+    },
     Database,
 };
+use socketioxide::{extract::SocketRef, SocketIo};
+use tokio::{signal, sync::mpsc};
+use tokio_util::{sync::CancellationToken, task::TaskTracker};
+use tower::ServiceBuilder;
 use tower_http::cors::{Any, CorsLayer};
+use tracing::{debug, info, level_filters::LevelFilter};
+use tracing_subscriber::{fmt::format::FmtSpan, EnvFilter};
 
 #[tokio::main]
 async fn main() {
-    let client: Database = Arc::new(PrismaClient::_builder().build().await.unwrap());
+    println!("Initializing scylla server...");
+    // construct a subscriber that prints formatted traces to stdout
+    // if RUST_LOG is not set, defaults to loglevel INFO
+    let subscriber = tracing_subscriber::fmt()
+        .pretty()
+        .with_thread_ids(true)
+        .with_ansi(true)
+        .with_thread_names(true)
+        .with_span_events(FmtSpan::CLOSE)
+        .with_env_filter(
+            EnvFilter::builder()
+                .with_default_directive(LevelFilter::INFO.into())
+                .from_env_lossy(),
+        )
+        .finish();
+    // use that subscriber to process traces emitted after this point
+    tracing::subscriber::set_global_default(subscriber).expect("Could not init tracing");
+
+    // create the database stuff
+    let db: Database = Arc::new(
+        PrismaClient::_builder()
+            .build()
+            .await
+            .expect("Could not build prisma DB"),
+    );
+
+    // create the socket stuff
+    let (socket_layer, io) = SocketIo::new_layer();
+    io.ns("/", |s: SocketRef| {
+        s.on_disconnect(|_: SocketRef| debug!("Socket: Client disconnected from socket"))
+    });
+
+    // channel to pass the mqtt data
+    // TODO tune buffer size
+    let (tx, rx) = mpsc::channel::<ClientData>(10000);
+
+    // channel to pass the processed data to the db thread
+    // TODO tune buffer size
+    let (tx_proc, rx_proc) = mpsc::channel::<Vec<ClientData>>(1000);
+
+    // the below two threads need to cancel cleanly to ensure all queued messages are sent.  therefore they are part of the a task tracker group.
+    // create a task tracker and cancellation token
+    let task_tracker = TaskTracker::new();
+    let token = CancellationToken::new();
+    // spawn the database handler
+    task_tracker.spawn(
+        db_handler::DbHandler::new(rx, Arc::clone(&db)).handling_loop(tx_proc, token.clone()),
```

Comment:
wouldnt cloning the token make a new token? therefore these are on two different tokens?

Reply:
You have to clone it to use it in the thread.  The clone keeps the link to the cancelation, tbh idk how it works but docs do say to clone.